Update rand to 0.9.x

dmitrmax opened the pull request:
Closes https://github.com/rust-bitcoin/rust-secp256k1/issues/789

Kixunil reviewed (approved):
ACK a9fe576a184591368fa317f05fb28f9fb5d72d91

Kixunil commented:
You need to run update lock files from contrib directory.

dmitrmax commented:
Done

apoelstra commented:
You also need to run the formatter and fix the compilation errors with the `rand` feature enabled.

Kixunil commented:
Also please squash the commits.

dmitrmax commented:
Ok, i did:
1) Fixed compilation errors
2) Squashed the commits
3) Run the rustfmt. But it formatted a lot of things that I've never touched myself. Is it ok?

dmitrmax commented:
I reverted rustfmt for the files which are not affected but transition to rand v0.9.0.  Only those touched by me are reformatted now.

apoelstra commented:
Seems like you are using the wrong version of the formatter. You need to use a nightly compiler to run `cargo fmt`.

dmitrmax commented:
Done

dmitrmax commented:
Fixed the no-std test

apoelstra commented:
The wasm failure looks like it's related to a call to `memmove` in libsecp. That isn't related to this PR. Similarly the Windows CI failure is hard to read but I think it's also unrelated. I'll run my local CI on this and see if it passes.

dmitrmax commented:
Ok. Let me know if there is anything more I can do for this PR.

apoelstra commented:
Ok, right now we depend on `getrandom 0.2` with the `js` feature enabled. This is done when compiling for wasm to turn this feature on because `rand` depends on it.

But `rand 0.9` does not depend on `getrandom 0.2`. It depends on `getrandom 0.3`. It appears that `getrandom 0.3` no longer has the `js` feature, even though the README for `rand 0.9` says it does https://crates.io/crates/rand

I'm not sure what exactly to do here, but we should definitely get rid of the `getrandom 0.2` dependency.

dmitrmax commented:
Ok, I've updated getrandom to 0.3. Somehow that required to update js-sys crate up to recent version 0.3.77. Sorry, I'm not a wasm guy,  I've called contrib/wasm.sh and it ran with the following error when executing wasm-bindgen:
`failed to find intrinsics to enable `clone_ref` function` 

Not sure what it means.

Also I've updated rand_core dependency from 0.6 to 0.9.

I've made a separate commit to simplify the review. If it looks good to you I can squash both commits into one.

apoelstra commented:
Can you squash? I think it's easiest to review this in one commit.

dmitrmax commented:
Squashed

apoelstra commented:
Thanks for iterating so much! This looks great now. Nice to move to `syn 2` across the board. Hopefully this works with our MSRV. CI will tell us.

I will also run my local CI on this which will do a bunch more checks, and then we should be good to merge.

apoelstra reviewed (approved):
ACK 993b4b75d77ce37f33c171504ede14d1e261af20; successfully ran local tests
